### PR TITLE
`iter(TensorProxy)` implementation

### PR DESCRIPTION
**Pull request by nikitaved**
Could that fix https://github.com/Lightning-AI/lightning-thunder/issues/717 based on https://github.com/Lightning-AI/lightning-thunder/issues/717#issuecomment-2210682835?

Not yet sure whether the whole approach is actually legit...

### CODE REVIEW

**Comment by t-vi** [comment]
I know I suggested lookaside, but as @kshitij12345 observed, it might be even easier to just add `TensorProxy.__iter__` directly (it will be run in the interpreter, so does the same as lookaside + impl that does it).
Maybe add a comment that this is typically run inside the Thunder interpreter.

**Review by t-vi — APPROVED** [review]
Thank you @nikitaved (and apologies for suggesting the more complicated initially).
Also thank you @kshitij12345 for the analysis and repro.

File: thunder/core/proxies.py [hunk 1]
```diff
@@ -1316,6 +1316,17 @@ def __getattr__(self, attr: str, /):
...
+    def __iter__(self):
+        # NOTE: this implementation is equivalent to torch.Tensor.__iter__
+
+        if self.ndim == 0:
+            raise TypeError("iteration over a 0-dim tensor")
+
+        import thunder.torch
+
+        unbound_tuple = thunder.torch.unbind(self)
+        return iter(unbound_tuple)
```
Comment (kshitij12345) [inline thunder/core/proxies.py:1328]: I think we can also do `iter(self.unbind()` as `unbind` is also a registered method.

https://github.com/Lightning-AI/lightning-thunder/blob/0d804443b2d499d897d0e5cb3cc96d508ec2454a/thunder/torch/__init__.py#L1256-L1262

Comment (nikitaved) [reply thunder/core/proxies.py:1328]: Hmm, did not know that things changed with the API :)

Comment (nikitaved) [reply thunder/core/proxies.py:1326]: Applied! Thank you!

**Pull request merged**